### PR TITLE
Add Share With Team and Make Private Menu Items to Stacks

### PR DESCRIPTION
**Pull request by hakankaradis**
Aim of this pr is giving ability to users to share (`Share With Team`) and unshare(`Make Private`) their private stacks with team.
## Description

Draft's sidebar getters updated to show shared stack on sidebar, instead of going stacks page and add to sidebar, now we are going to show all drafts on sidebar by default.

Draft's stack's menu options updated. Shared stack templates menu options came with `Edit` option but now it is `View Stack`

Show `Share With Team` and `Make Private` menu options for stacks.

Remove Validator.own stack template permission to generate stack from shared stack template.
## Bugs
- [x] Updates came by refresh when shared stack template has update
- [x] If user delete shared stack template, others will have no longer that stack template.
## Motivation and Context
#9225 
## Screenshots (if appropriate):

http://recordit.co/WpUjt58R9g
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

**Comment by gokmen** [comment]
@hakankaradis ping for bugs mentioned in PR description.

**Comment by hakankaradis** [comment]
@gokmen I will update you soon, didn't have time to work on it.

**Comment by hakankaradis** [comment]
- Update SendNotification func signature and add sharedStackTemplates attribute.
- ComputeUtils: fetchGroupStackTemplate update func signature
- ComputeProvider: createGroupStack update func signature
- groupController: update setDefaultTemplate (accept type) to use it for share with team func.

Current group will have all shared stackTemplates ( in stackTemplates and sharedStackTemplates)

- sharedStackTemplates is an array and hold only default stack template id.
- stackTemplates is an array and hold all shared stack template ids.

Currently having problem when user unshare his team stack.
I will update soon. @gokmen

**Comment by hakankaradis** [comment]
@gokmen here is final screencast, I will push my commits soon. http://recordit.co/EcWgWfFBf6

**Comment by gokmen** [comment]
@hakankaradis can u fix tests and re-assign me back, thanks.

**Comment by hakankaradis** [comment]
@gokmen 
I added another control flow for destroying team or shared stack.
Here are the screenshots for admins and members

File: client/app/lib/components/sidebarstacksection/index.coffee [hunk 1]
```diff
@@ -106,6 +108,12 @@ module.exports = class SidebarStackSection extends React.Component
       when 'Make Team Default'
         remote.api.JStackTemplate.one { _id: templateId }, (err, template) ->
           computeController.makeTeamDefault template, no  unless err
+      when 'Share Stack With Team'
+        remote.api.JStackTemplate.one { _id: templateId }, (err, template) ->
+          computeController.shareStackWithTeam template  unless err
+      when 'Make Stack Private'
+        remote.api.JStackTemplate.one { _id: templateId }, (err, template) ->
+          computeController.makeStackPrivate template  unless err
```
Comment (gokmen) [inline client/app/lib/components/sidebarstacksection/index.coffee]: let's move these functions from computeController to JStackTemplate's remote extension and here we can say `template.makePrivate()` and `template.shareStackWithTeam()`

File: client/app/lib/providers/computecontroller.coffee [hunk 2]
```diff
@@ -1203,22 +1213,56 @@ module.exports = class ComputeController extends KDController
...
+
+    createShareModal (needShare, modal) =>
+
+      groupsController.setDefaultTemplate stackTemplate, 'share', (err) =>
+        if needShare
+        then @shareCredentials credentials, requiredProviders, -> modal.destroy()
+        else modal.destroy()
+
+
+  makeStackPrivate: (stackTemplate) ->
```
Comment (gokmen) [inline client/app/lib/providers/computecontroller.coffee]: this shouldn't have a UI code here, just the functionality should be here. If you need to provide some UI helper you can put them into `ComputController.UI`

File: workers/social/lib/social/models/computeproviders/computeutils.coffee [hunk 3]
```diff
@@ -273,12 +273,15 @@ fetchGroupStackTemplate = (client, callback) ->
       else
         return callback new KodingError 'Template not set', 'NotFound'
 
+    { _id } = options
     # TODO Make this works with multiple stacks ~ gg
-    stackTemplateId = group.stackTemplates[0]
+
+    _id = if _id then _id else group.sharedStackTemplates?[0] or group.stackTemplates[0]
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/computeutils.coffee]: `_id ?= ...`

File: workers/social/lib/social/models/computeproviders/computeutils.coffee [hunk 4]
```diff
@@ -273,12 +273,15 @@ fetchGroupStackTemplate = (client, callback) ->
...
 
     # TODO make all these in seperate functions
     JStackTemplate = require './stacktemplate'
-    JStackTemplate.one { _id: stackTemplateId }, (err, template) ->
+
+    JStackTemplate.one { _id: _id }, (err, template) ->
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/computeutils.coffee]: `{ _id }`

**Review by gokmen — CHANGES_REQUESTED** [review]
There will be more comments on this one but can u update the ones commented first, thanks!

**Comment by hakankaradis** [comment]
@gokmen  http://recordit.co/8vnFejfVIp

File: client/app/lib/components/sidebar/index.coffee [hunk 5]
```diff
@@ -92,22 +93,35 @@ module.exports = class Sidebar extends React.Component
...
       when 'Initialize'
         EnvironmentFlux.actions.generateStack(id).then ({ template }) ->
           appManager.tell 'Stackeditor', 'reloadEditor', template._id
       when 'Clone'
-        EnvironmentFlux.actions.cloneStackTemplate draft.toJS(), yes
+        remote.api.JStackTemplate.one { _id }, (err, template) =>
+          return @showErrorMessage 'Error occured while cloning template'  if err
+          EnvironmentFlux.actions.cloneStackTemplate template, (err) =>
+            return @showErrorMessage 'Error occured while cloning template'  if err
+            return @showErrorMessage 'Your stack template successfully cloned'
```
Comment (gokmen) [inline client/app/lib/components/sidebar/index.coffee]: it looks weird that we're showing a success message over `showErrorMessage` method.

File: client/app/lib/components/sidebar/index.coffee [hunk 6]
```diff
@@ -92,22 +93,35 @@ module.exports = class Sidebar extends React.Component
...
+        remote.api.JStackTemplate.one { _id }, (err, template) =>
+          return @showErrorMessage 'Error occured while making this stack team default'  if err
+          EnvironmentFlux.actions.makeTeamDefault template, (err) =>
+            return @showErrorMessage 'Error occured while making this stack team default'  if err
+            return @showErrorMessage 'Your stack is team default now.'
       when 'Delete'
         computeController.deleteStackTemplate draft.toJS(), yes
 
 
+  showErrorMessage: (title) -> new kd.NotificationView { title }
```
Comment (gokmen) [inline client/app/lib/components/sidebar/index.coffee]: better to rename this as `showMessage`

File: client/app/lib/components/sidebarstacksection/index.coffee [hunk 7]
```diff
@@ -84,28 +86,48 @@ module.exports = class SidebarStackSection extends React.Component
...
+            return @showErrorMessage 'Your stack successfully shared with team'
+      when 'Make Stack Private'
+        remote.api.JStackTemplate.one { _id }, (err, template) =>
+          return @showErrorMessage 'Error occured while unsharing this stack'  if err
+          EnvironmentFlux.actions.makePrivate template, (err) =>
+            return @showErrorMessage 'Error occured while unsharing this stack'  if err
+            return @showErrorMessage 'Your stack is private now.'
+
+
+  showErrorMessage: (title) -> new kd.NotificationView { title }
```
Comment (gokmen) [inline client/app/lib/components/sidebarstacksection/index.coffee]: same issue here.

File: client/app/lib/components/sidebarstacksection/index.coffee [hunk 8]
```diff
@@ -211,5 +243,7 @@ module.exports = class SidebarStackSection extends React.Component
       {@renderStackUpdatedWidget()}
     </SidebarSection>
 
+isStackMine = (stackOriginId, myId) -> stackOriginId is myId
```
Comment (gokmen) [inline client/app/lib/components/sidebarstacksection/index.coffee]: no one is using this it seems.

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee [hunk 9]
```diff
@@ -337,6 +423,42 @@ module.exports = class JStackTemplate extends Module
...
+
+      if stackTemplates.length
+        unless id in stackTemplates
+          stackTemplates.push id
+      else stackTemplates = [id]
+
+      sharedStackTemplates = [id]
+
+      @setAccess client, 'group', (err) =>
+        group.modify client, { stackTemplates, sharedStackTemplates }, (err) =>
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: why do we need that? and why we are keeping this list on `JGroup` instance?
we can get list of templates shared with group by simply `JStackTemplate.some {accessLevel: 'group'}`

Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: and we shouldn't modify such data from this place at all.

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee [hunk 10]
```diff
@@ -322,7 +339,76 @@ module.exports = class JStackTemplate extends Module
...
+      stackTemplates = makeArray stackTemplates
+      id = @getId().toHexString()
+
+      if stackTemplates.length
+        index = stackTemplates.indexOf(id)
+        if index > -1
+          stackTemplates.splice(index, 1)
+
+      @setAccess client, 'private', (err) =>
+        group.modify client, { stackTemplates }, (err) =>
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: same here, `JGroup` instance data shouldn't be modified by another model such as `JStackTemplate`

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee [hunk 11]
```diff
@@ -322,7 +339,76 @@ module.exports = class JStackTemplate extends Module
...
+        unless id in stackTemplates
+          stackTemplates.push id
+      else stackTemplates = [id]
+
+      console.log 'stackTemplates', stackTemplates
+      @setAccess client, 'group', (err) =>
+        group.modify client, { stackTemplates }, (err) =>
+          return callback err  if err
+
+          return callback err, this
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: we don't need this method at all, you can simply use `JStackTemplate::setAccess 'group'` from client side. same goes for `makePrivate` method as well.

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee [hunk 12]
```diff
@@ -322,7 +339,76 @@ module.exports = class JStackTemplate extends Module
...
+          stackTemplates.splice(index, 1)
+
+      @setAccess client, 'private', (err) =>
+        group.modify client, { stackTemplates }, (err) =>
+          return callback err  if err
+
+          return callback err, this
+
+
+  makeTeamDefault: permit
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: this is not a feature of `JStackTemplate` it's a feature of `JGroup` and it already has a functionality for this https://github.com/koding/koding/blob/master/workers/social/lib/social/models/group/index.coffee#L1108-L1136

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee [hunk 13]
```diff
@@ -337,6 +423,42 @@ module.exports = class JStackTemplate extends Module
...
+      { permission: 'update own stack template' }
+      { permission: 'update stack template' }
+    ]
+
+    success: revive
+
+      shouldReviveClient   : yes
+      shouldReviveProvider : no
+      shouldFetchGroupLimit : yes
+
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: we don't need this change as well.

File: workers/social/lib/social/models/group/index.coffee [hunk 14]
```diff
@@ -289,6 +291,8 @@ module.exports = class JGroup extends Module
       # it would break the onboarding process of showing the initial stacks not
       # configured modal, so don't. - SY
       stackTemplates  : [ ObjectId ]
+
+      sharedStackTemplates: [ ObjectId ]
```
Comment (gokmen) [inline workers/social/lib/social/models/group/index.coffee]: let's revert this also.

File: workers/social/lib/social/models/computeproviders/stacktemplate.coffee [hunk 15]
```diff
@@ -65,6 +65,13 @@ module.exports = class JStackTemplate extends Module
           (signature String, Function)
         hasStacks     :
           (signature Function)
+        shareWithTeam :
+          (signature Function)
+        makePrivate   :
+          (signature Function)
+        makeTeamDefault:
+          (signature Function)
```
Comment (gokmen) [inline workers/social/lib/social/models/computeproviders/stacktemplate.coffee]: let's revert this, details are in comments.

**Review by gokmen — CHANGES_REQUESTED** [review]
we may need a session for this, it shouldn't be handled in this way. I added some more comments but if you need further information please ping me before taking action.

**Comment by gokmen** [comment]
@hakankaradis closing this one, when you're ready we can go over with the new one.

**Pull request closed without merging**